platform/distance.cpp

```diff
@@ -0,0 +1,246 @@
+  case Distance::Units::Kilometers:
+    return Distance(measurement_utils::MilesToMeters(distance) / 1000, Distance::Units::Kilometers);
+  case Distance::Units::Feet:
+    return Distance(measurement_utils::MilesToFeet(distance), Distance::Units::Feet);
+  default: UNREACHABLE();
+  }
+}
+
+Distance NauticalMilesTo(double distance, Distance::Units units)
+{
+  switch (units)
+  {
+  case Distance::Units::Meters:
+    return Distance(measurement_utils::NauticalMilesToMeters(distance), Distance::Units::Meters);
+  default: UNREACHABLE();
+  }
+}
+
+Distance InchesTo(double distance, Distance::Units units)
+{
+  switch (units)
+  {
+  case Distance::Units::Meters:
+    return Distance(measurement_utils::InchesToMeters(distance), Distance::Units::Meters);
+  default: UNREACHABLE();
+  }
+}
+
+double WithPrecision(double value, uint8_t precision)
+{
+  std::ostringstream ss;
+  ss << std::setprecision(precision) << std::fixed << value;
+  return std::atof(ss.str().c_str());
+}
+
+}  // namespace
+
+Distance::Distance() : Distance(0.0) {}
+
+Distance::Distance(double distanceInMeters) : Distance(distanceInMeters, Units::Meters) {}
+
+Distance::Distance(double distance, platform::Distance::Units units)
+  : m_distance(distance), m_units(units)
+{
+}
+
+Distance Distance::CreateFormatted(double distanceInMeters)
+{
+  return Distance(distanceInMeters).ToPlatformUnitsFormatted();
+}
+
+bool Distance::IsValid() const { return m_distance > std::numeric_limits<double>::epsilon(); }
+
+Distance Distance::To(Distance::Units units) const
+{
+  if (m_units == units)
+    return *this;
+
+  switch (m_units)
+  {
+  case Units::Meters: return MetersTo(m_distance, units);
+  case Units::Kilometers: return KilometersTo(m_distance, units);
+  case Units::Feet: return FeetTo(m_distance, units);
+  case Units::Miles: return MilesTo(m_distance, units);
+  case Units::NauticalMiles: return NauticalMilesTo(m_distance, units);
+  case Units::Inches: return InchesTo(m_distance, units);
+  default: UNREACHABLE();
+  }
+}
+
+Distance Distance::ToPlatformUnitsFormatted() const
+{
+  return To(measurement_utils::GetMeasurementUnits() == measurement_utils::Units::Metric
+                ? Units::Meters
+                : Units::Feet)
+      .GetFormattedDistance();
+}
+
+double Distance::GetDistance() const { return m_distance; }
+
+Distance::Units Distance::GetUnits() const { return m_units; }
+
+std::string Distance::GetDistanceString() const
+{
+  std::ostringstream os;
+  os << std::defaultfloat << m_distance;
+  return os.str();
+}
+
+// LocalizedStrings are supported only on iOS (Android is broken)
+#if defined(OMIM_OS_IPHONE)
+std::string Distance::GetUnitsString() const
+{
+  ASSERT_NOT_EQUAL(m_units, Units::NauticalMiles, ("OSM-only units are not supported"));
+  ASSERT_NOT_EQUAL(m_units, Units::Inches, ("OSM-only units are not supported"));
+
+  switch (m_units)
+  {
+  case Units::Meters: return GetLocalizedString("meter");
```

Comment:
Is there an issue for it on Android? Did you see why is it broken?

Reply:
That's the first reason of this PR 😁 
No, it's not working. Why? Idk.
Anyway, there is no sense to call this function from Android:
* In cpp call Java to return string
* Put it in std::string
* Convert it to jstring
* Pass to java

It's better to get string in java directly

Reply:
Hm... That's weird but in some cases it works 🤔